doc/developer/design/20210505_kafka_keys_in_sql.md

```diff
@@ -0,0 +1,302 @@
+// If applicable, be sure to call out any new testing/validation that will be required
+-->
+
+### Syntax and user-visible semantics
+
+Enhance the Kafka create source syntax like so:
+
+```sql
+CREATE SOURCE <source-name> FROM KAFKA BROKER '<>'
+<format>
+[INCLUDE KEY [AS <key-column-name>]]
+```
+
+where `<format>` is either:
+
+* An explicit key/value format:
+
+  ```sql
+  KEY FORMAT <specifier>
+  VALUE FORMAT <specifier>
+  ```
+
+* A bare format that specifies a Confluent Schema Registry (CSR) config
+
+  ```sql
+  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY ...
+  ```
+
+If `INCLUDE KEY` is specified, the source schema will be augmented to prepend
+all of the values of the key as columns in the dataflow. Format types that do
+not specify subfields (e.g. `FORMAT TEXT`) will use their standard column names
+(i.e. `text`).
+
+Alternatively, if the `INCLUDE KEY AS <name>` syntax is used then the key value
+will get the name specified:
+
+* If the field type is simple (bytes or text) the field will just be renamed
+* If the field type has defined fields, the fields will be included as a record
+  with that name.
+
+If any key column names colide with value columns using the bare `INCLUDE KEY`
+syntax, an error will be raised before dataflow construction suggesting the use
+of the `AS <name>` clause.
+
+Justification for this set of defaults is that it is reasonable to expect that
+the common case for the `INCLUDE KEY` will not have collisions between the key
+and value fields: if the key fields are a subset of the value fields (as is the
+case in Debezium) then there isn't any reason to include the key, since all its
+data is already in the value.
+
+#### Usage Examples
+
+##### Text/Text
+
+```sql
+CREATE SOURCE text_text FROM KAFKA BROKER '...' TOPIC '...'
+KEY FORMAT TEXT VALUE FORMAT TEXT
+INCLUDE KEY AS key
+ENVELOPE UPSERT;
+```
+
+<details>
+<summary>👨‍🔬 It is possible to express the above with a bare format. 👨‍🔬</summary>
+
+```sql
+CREATE SOURCE text_text FROM KAFKA BROKER '...' TOPIC '...'
+FORMAT TEXT
+INCLUDE KEY AS key
+ENVELOPE UPSERT;
+```
+
+</details>
+
+This will create a new source, it requires specifying the record syntax because
+we use the same name by default, which would result in a name conflict.
+
+Usage in a view looks like:
+
+```sql
+CREATE MATERIALIZED VIEW text_text_view AS
+SELECT key, text FROM text_text
+WHERE ...;
+```
+
+##### Text/Avro
+
+```sql
+CREATE SOURCE text_avro FROM KAFKA BROKER '...' TOPIC '...'
+KEY FORMAT TEXT
+VALUE FORMAT AVRO USING SCHEMA '{"type": "record", "name": "value", "fields": [ {"name": "field", "type": "int"} ] }';
+```
+
+Usage in a view looks like:
+
+```sql
+CREATE MATERIALIZED VIEW text_avro_view AS
+SELECT text, field as interesting FROM text_avro;
+```
+
+##### Avro/Avro
+
+```sql
+CREATE SOURCE avro_avro FROM KAFKA BROKER '...' TOPIC '...'
+KEY FORMAT AVRO USING SCHEMA '{"type": "record", "name": "boring", "fields": [
```

Comment:
Did you intentionally not write `INCLUDE KEY` in this example?

Reply:
nope, don't know how that happened.